crates/matrix-sdk/src/authentication/oauth/mod.rs

```diff
@@ -1518,12 +1523,127 @@ impl<'a> LoginWithQrCodeBuilder<'a> {
 pub struct GrantLoginWithQrCodeBuilder<'a> {
     /// The underlying Matrix API client.
     client: &'a Client,
+    /// The duration to wait for the homeserver to create the new device after
+    /// consenting the login before giving up.
+    device_creation_timeout: Duration,
 }
 
 #[cfg(feature = "e2e-encryption")]
 impl<'a> GrantLoginWithQrCodeBuilder<'a> {
-    /// Grant login by generating a QR code on this device to be scanned by the
-    /// new device.
+    /// Create a new builder with the default device creation timeout.
+    pub fn new(client: &'a Client) -> Self {
+        Self { client, device_creation_timeout: Duration::from_secs(10) }
+    }
+
+    /// Set the device creation timeout.
+    ///
+    /// # Arguments
+    ///
+    /// * `device_creation_timeout` - The duration to wait for the homeserver to
+    ///   create the new device after consenting the login before giving up.
+    pub fn device_creation_timeout(mut self, device_creation_timeout: Duration) -> Self {
+        self.device_creation_timeout = device_creation_timeout;
+        self
+    }
+
+    /// This method allows you to grant login to a new device by scanning a
+    /// QR code generated by the new device.
+    ///
+    /// The new device needs to display the QR code which this device can
+    /// scan and call this method to grant the login.
+    ///
+    /// A successful login grant using this method will automatically mark the
+    /// new device as verified and transfer all end-to-end encryption
+    /// related secrets, like the private cross-signing keys and the backup
+    /// key from this device device to the new device.
+    ///
+    /// For the reverse flow where this device generates the QR code
+    /// for the new device to scan, use
+    /// [`GrantLoginWithQrCodeBuilder::generate`].
+    ///
+    /// # Arguments
+    ///
+    /// * `data` - The data scanned from a QR code.
+    ///
+    /// # Example
+    ///
+    /// ```no_run
+    /// use anyhow::bail;
+    /// use futures_util::StreamExt;
+    /// use matrix_sdk::{
+    ///     Client, authentication::oauth::{
+    ///         qrcode::{GrantLoginProgress, QrCodeData, QrCodeModeData, QrProgress},
+    ///     }
+    /// };
+    /// use std::{error::Error, io::stdin};
+    /// # _ = async {
+    /// # let bytes = unimplemented!();
+    /// // You'll need to use a different library to scan and extract the raw bytes from the QR
+    /// // code.
+    /// let qr_code_data = QrCodeData::from_bytes(bytes)?;
+    ///
+    /// // Fetch the homeserver out of the parsed QR code data.
+    /// let QrCodeModeData::Reciprocate{ server_name } = qr_code_data.mode_data else {
+    ///     bail!("The QR code is invalid, we did not receive a homeserver in the QR code.");
+    /// };
```

Comment:
Does this make sense in this example? We're the already logged in client which is granting a log in by scanning a QR code.

We're already using a homeserver, no?